include/kernels/dslash_pack.cuh

```diff
@@ -360,48 +548,115 @@ namespace quda
-      case 0:
-        while (local_tid < arg.nFace * arg.dc.ghostFaceCB[0]) {
-          int ghost_idx = dir * arg.nFace * arg.dc.ghostFaceCB[0] + local_tid;
-          if (arg.nFace == 1)
-            packStaggered<0, 1>(arg, ghost_idx, s, parity);
-          else
-            packStaggered<0, 3>(arg, ghost_idx, s, parity);
-          local_tid += arg.blocks_per_dir * blockDim.x;
+      // for shmem
+      const int shmem = arg.shmem;
+      const int shmemidx = 2 * dim + dir;
+      bool intranode = getShmemBuffer<1, decltype(arg)>(shmemidx, arg) == nullptr;
+      bool pack_intranode = (!arg.packkernel) != (!(shmem & 1));
+      bool pack_internode = (!arg.packkernel) != (!(shmem & 2));
+      if (shmem == 0 || (intranode && pack_intranode) || (!intranode && pack_internode)) {
+        switch (dim) {
+        case 0:
+          while (local_tid < arg.nFace * arg.dc.ghostFaceCB[0]) {
+            int ghost_idx = dir * arg.nFace * arg.dc.ghostFaceCB[0] + local_tid;
+            if (arg.nFace == 1)
+              packStaggered<0, 1>(arg, ghost_idx, s, parity);
+            else
+              packStaggered<0, 3>(arg, ghost_idx, s, parity);
+            local_tid += arg.blocks_per_dir * blockDim.x;
+          }
+          break;
+        case 1:
+          while (local_tid < arg.nFace * arg.dc.ghostFaceCB[1]) {
+            int ghost_idx = dir * arg.nFace * arg.dc.ghostFaceCB[1] + local_tid;
+            if (arg.nFace == 1)
+              packStaggered<1, 1>(arg, ghost_idx, s, parity);
+            else
+              packStaggered<1, 3>(arg, ghost_idx, s, parity);
+            local_tid += arg.blocks_per_dir * blockDim.x;
+          }
+          break;
+        case 2:
+          while (local_tid < arg.nFace * arg.dc.ghostFaceCB[2]) {
+            int ghost_idx = dir * arg.nFace * arg.dc.ghostFaceCB[2] + local_tid;
+            if (arg.nFace == 1)
+              packStaggered<2, 1>(arg, ghost_idx, s, parity);
+            else
+              packStaggered<2, 3>(arg, ghost_idx, s, parity);
+            local_tid += arg.blocks_per_dir * blockDim.x;
+          }
+          break;
+        case 3:
+          while (local_tid < arg.nFace * arg.dc.ghostFaceCB[3]) {
+            int ghost_idx = dir * arg.nFace * arg.dc.ghostFaceCB[3] + local_tid;
+            if (arg.nFace == 1)
+              packStaggered<3, 1>(arg, ghost_idx, s, parity);
+            else
+              packStaggered<3, 3>(arg, ghost_idx, s, parity);
+            local_tid += arg.blocks_per_dir * blockDim.x;
+          }
+          break;
         }
-        break;
-      case 1:
-        while (local_tid < arg.nFace * arg.dc.ghostFaceCB[1]) {
-          int ghost_idx = dir * arg.nFace * arg.dc.ghostFaceCB[1] + local_tid;
-          if (arg.nFace == 1)
-            packStaggered<1, 1>(arg, ghost_idx, s, parity);
-          else
-            packStaggered<1, 3>(arg, ghost_idx, s, parity);
-          local_tid += arg.blocks_per_dir * blockDim.x;
+      }
+
+#ifdef NVSHMEM_COMMS
```

Comment:
This is the same code as above for the non-staggered kernel.  Can you move this to its own function to avoid the repeated boilerplate?

Reply:
Did that refer to the shmem_signal you already extracted?

Reply:
Yep